proposals/NNNN-spirv-types-in-clang.md

```diff
@@ -0,0 +1,113 @@
+`SpirvType` specializations while lowering types. However, this approach does
+not interact well with Sema and the rest of Clang, resulting in at least several
+bugs in handling template arguments and other areas. In order to take the same
+approach in Clang, we would need to add special case handling while lowering
+struct types, as well as for calculating size and alignment and anywhere else
+structs are interacted with in Clang. Struct types are already processed
+separately from all other types while lowering, and adding special cases for
+`SpirvType` would complicate their handling even more.
+
+Finally, a struct is not a good semantic representation of a `SpirvType`.
+`SpirvType` is fundamentally a new kind of type for Clang that is different from
+all existing types—it represents a low-level target type specified by the user.
+
+## Proposed solution
+
+### Type Representation
+
+Describe your solution to the problem. Provide examples and describe how they
+work. Show how your solution is better than current workarounds: is it cleaner,
+safer, or more efficient?
+
+`HLSLInlineSpirvType` is a subclass of `clang::Type`, which has the properties
+
+| Property    | Optional | Type                   |
+| ----------- | -------- | ---------------------- |
+| `Operand`   | no       | `uint32_t`             |
+| `Size`      | yes      | `uint32_t`             |
+| `Alignment` | yes      | `uint32_t`             |
+| `Operands`  | no       | array of`SpirvOperand` |
+
+A value of type `HLSLInlineSpirvType` is always canonical.
+
+#### Operands
+
+Each operand in `Operands` is a value of the nested class
+`HLSLInlineSpirvType::SpirvOperand`, which can be one of the following kinds:
+
+| Kind         | Description                                                                        | Has Result Type | Has Integral Value |
+| ------------ | ---------------------------------------------------------------------------------- | --------------- | ------------------ |
+| `ConstantId` | Represents a value to be passed in as the ID of a SPIR-V `OpConstant` instruction. | yes             | yes                |
+| `Literal`    | Represents a value to be passed in as an immediate literal.                        | no              | yes                |
+| `TypeId`     | Represents a type to be passed in as the ID of a SPIR-V `OpType*` instruction      | yes             | no                 |
+
+### Type Declaration
+
+Since `HLSLInlineSpirvType` types should only be created using the `SpirvType`
+and `SpirvOpaqueType` templates, we do not need to provide special syntax for
+their declaration. In order to implement these templates, and to avoid the
+necessity of creating an additional dependent type, these types will be created
+using a `vk::__hlsl_spirv_type` builtin template.
+
+```C++
+template <uint32_t Opcode, uint32_t Size, uint32_t Alignment,
+          typename... Operands>
+using __hlsl_spirv_type = ...;
```

Comment:
Is this something that will appear in the header file or is this implemented as if that was in the source code?

I'm not clear on what the `...` will be. How will this link the builtin type with `SpirvType` and `SpirvOpaqueType`?

Reply:
I've added more explanation in this section, including example declarations of `SpirvType` and `SpirvOpaqueType`.